### PR TITLE
Gracefully handle error conditions

### PR DESCRIPTION
**Pull request by DenverM80**
CSDK-43: headObject returns head bucket if passed empty object
CSDK-44: getBucket returns getBuckets value if passed empty string

==22280== LEAK SUMMARY:
==22280==    definitely lost: 0 bytes in 0 blocks
==22280==    indirectly lost: 0 bytes in 0 blocks
==22280==      possibly lost: 0 bytes in 0 blocks
==22280==    still reachable: 2,550 bytes in 16 blocks
==22280==         suppressed: 0 bytes in 0 blocks
==22280== 
==22280== For counts of detected and suppressed errors, rerun with: -v
==22280== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
denverm@dm-lt:~/code/github/ds3_c_sdk/test$

### CODE REVIEW

File: src/ds3.c [hunk 1]
```diff
@@ -1090,12 +1090,17 @@ static ds3_str* _build_path(const char* path_prefix, const char* bucket_name, co
     }
 
     joined_path = g_strjoin("/", escaped_bucket_name, escaped_object_name, NULL);
-    full_path = g_strconcat(path_prefix, joined_path, NULL);
+    if( TRUE == g_str_has_suffix(joined_path, "/")) {
```
Comment (rpmoore) [inline src/ds3.c]: It looks like the space next to the `(` is on the wrong side.

**Comment by rpmoore** [comment]
How about a test that has some bucket names with trailing slashes?

**Comment by DenverM80** [comment]
will update with another unit test

**Comment by DenverM80** [comment]
Ready for re-review

File: src/ds3.c [hunk 2]
```diff
@@ -1766,10 +1777,27 @@ ds3_error* ds3_get_bucket(const ds3_client* client, const ds3_request* request,
     return NULL;
 }
 
+static int num_chars_in_ds3_str( ds3_str* str, char ch) {
```
Comment (rpmoore) [inline src/ds3.c]: You have an extra space between the `(` and `ds3_str`.  Also, can ds3_str be `const` here?

File: src/ds3.c [hunk 3]
```diff
@@ -1695,6 +1701,11 @@ ds3_error* ds3_get_bucket(const ds3_client* client, const ds3_request* request,
     ds3_error* error;
     GArray* object_array;
     GArray* common_prefix_array;
+
+    if ( 0 == g_strcmp0(request->path->value, "/") ){
```
Comment (rpmoore) [inline src/ds3.c]: There are extra spaces on the inside of the parenthesis.

File: test/bucket_tests.cpp [hunk 4]
```diff
@@ -152,6 +153,33 @@ BOOST_AUTO_TEST_CASE(max_keys) {
     free_client(client);
 }
 
+
+//testing get_bucket with bucket_name/ with trailing slash
+BOOST_AUTO_TEST_CASE( put_bucket_bucket_name_with_trailing_slash){
```
Comment (rpmoore) [inline test/bucket_tests.cpp]: There is an extra space next to the `(`

File: test/bucket_tests.cpp [hunk 5]
```diff
@@ -152,6 +153,33 @@ BOOST_AUTO_TEST_CASE(max_keys) {
...
+    handle_error(error);
+
+    request = ds3_init_get_bucket(bucket_name);
+    ds3_get_bucket_response* response;
+    error = ds3_get_bucket(client, request, &response);
+    printf("bucket_name[%s]\n", response->name->value);
+    ds3_free_request(request);
+
+    BOOST_CHECK(error == NULL);
+    BOOST_CHECK_EQUAL( g_strcmp0(response->name->value, "trailing_slash"), 0);
```
Comment (rpmoore) [inline test/bucket_tests.cpp]: There is an extra space next to the `(`

**Pull request merged**